pkg/http/api.go

```diff
@@ -1682,60 +1684,87 @@ func (a *api) onDirectMessage(reqCtx *fasthttp.RequestCtx) {
 			if status.Code(rErr) == codes.PermissionDenied {
 				invokeErr.statusCode = invokev1.HTTPStatusFromCode(codes.PermissionDenied)
 			}
-			return dmpr, invokeErr
-		}
-
-		dmpr.headers = rResp.Headers()
-		dmpr.contentType = rResp.ContentType()
-		dmpr.body, rErr = rResp.RawDataFull()
-		if rErr != nil {
-			return dmpr, invokeError{
-				statusCode: fasthttp.StatusInternalServerError,
-				msg:        NewErrorResponse("ERR_DIRECT_INVOKE", fmt.Sprintf(messages.ErrDirectInvoke, targetID, rErr)),
-			}
+			return rResp, invokeErr
 		}
 
-		// Construct response
-		dmpr.statusCode = int(rResp.Status().Code)
+		// Construct response if not HTTP
+		resStatus := rResp.Status()
 		if !rResp.IsHTTPResponse() {
-			dmpr.statusCode = invokev1.HTTPStatusFromCode(codes.Code(dmpr.statusCode))
-			if dmpr.statusCode != fasthttp.StatusOK {
-				if dmpr.body, rErr = invokev1.ProtobufToJSON(rResp.Status()); rErr != nil {
-					return dmpr, invokeError{
+			statusCode := int32(invokev1.HTTPStatusFromCode(codes.Code(resStatus.Code)))
+			if statusCode != fasthttp.StatusOK {
+				// Close the response to replace the body
+				_ = rResp.Close()
+				var body []byte
+				body, rErr = invokev1.ProtobufToJSON(resStatus)
+				rResp.WithRawDataBytes(body)
+				resStatus.Code = statusCode
+				if rErr != nil {
+					return rResp, invokeError{
 						statusCode: fasthttp.StatusInternalServerError,
 						msg:        NewErrorResponse("ERR_MALFORMED_RESPONSE", rErr.Error()),
 					}
 				}
+			} else {
+				resStatus.Code = statusCode
 			}
-		} else if dmpr.statusCode != fasthttp.StatusOK {
-			return dmpr, fmt.Errorf("Received non-successful status code: %d", dmpr.statusCode) //nolint:stylecheck
+		} else if resStatus.Code != fasthttp.StatusOK {
+			return rResp, fmt.Errorf("Received non-successful status code: %d", resStatus.Code) //nolint:stylecheck
 		}
-		return dmpr, nil
+		return rResp, nil
 	})
 
 	// Special case for timeouts/circuit breakers since they won't go through the rest of the logic.
 	if errors.Is(err, context.DeadlineExceeded) || breaker.IsErrorPermanent(err) {
 		respond(reqCtx, withError(fasthttp.StatusInternalServerError, NewErrorResponse("ERR_DIRECT_INVOKE", err.Error())))
+		cancel()
 		return
 	}
 
-	if dmpr != nil && len(dmpr.headers) > 0 {
-		invokev1.InternalMetadataToHTTPHeader(reqCtx, dmpr.headers, reqCtx.Response.Header.Set)
+	if resp != nil {
```

Comment:
Shouldn't headers be set even if the response is empty?

Reply:
Addressed on Teams